File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/SliderPanel.java
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.ui;

import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.uiDesigner.core.GridConstraints;
import com.intellij.uiDesigner.core.GridLayoutManager;
import java.awt.Dimension;
import java.awt.Insets;
import javax.swing.JComponent;
import org.jetbrains.annotations.TestOnly;

import javax.swing.JPanel;
import javax.swing.JSlider;
import javax.swing.JTextField;
import java.awt.event.FocusAdapter;
import java.awt.event.FocusEvent;

import static software.aws.toolkits.resources.Localization.message;

// A panel with a slider and text field, of which the slider and text field always synced up.
public class SliderPanel {
    private final int min, max, minTick, maxTick, minorTick, majorTick, defaultValue;
    private final boolean snap;

    private JPanel content;
    public JSlider slider;
    public JTextField textField;

    public SliderPanel(int min, int max, int defaultValue, int minTick, int maxTick, int minorTick, int majorTick, boolean snap) {
        this.min = min;
        this.max = max;
        this.defaultValue = defaultValue;
        this.minTick = minTick;
        this.maxTick = maxTick;
        this.minorTick = minorTick;
        this.majorTick = majorTick;
        this.snap = snap;
        bind();
    }

    public void setValue(int value) {
        slider.setValue(value);
    }

    public int getValue() {
        return slider.getValue();
    }

    public ValidationInfo validate() {
        Integer value = null;
        try {
            value = Integer.parseInt(textField.getText());
        } catch (Exception ignored) {
        }

        if (value == null || value < min || value > max) {
            return new ValidationInfo(message("lambda.slider_validation", min, max), textField);
        }
        return null;
    }

    public void setEnabled(Boolean enabled) {
        slider.setEnabled(enabled);
        textField.setEnabled(enabled);
    }

    @TestOnly
    public boolean isVisible() {
        return slider.getParent().isVisible() && slider.isVisible() && textField.isVisible();
    }

    private void bind() {
        slider.setMajorTickSpacing(majorTick);
        slider.setMinorTickSpacing(minorTick);
        slider.setMaximum(maxTick);
        slider.setMinimum(minTick);
        slider.setPaintLabels(true);
        slider.setSnapToTicks(snap);
        slider.setValue(defaultValue);
        slider.addChangeListener(e ->
                                     textField.setText(Integer.toString(validValue(slider.getValue())))
        );
        textField.setText(Integer.toString(slider.getValue()));
        textField.addFocusListener(new FocusAdapter() {
            // When the text field lost focus, we force the value to be valid to reset to
            // - default value if the input is not a valid integer, or
            // - min if it is smaller than min, or
            // - max if it is bigger than max.
            @Override
            public void focusLost(FocusEvent e) {
                int value;
                try {
                    value = validValue(Integer.parseInt(textField.getText()));
                } catch (Exception e2) {
                    value = defaultValue;
                }
                slider.setValue(value);
                textField.setText(Integer.toString(value));
            }
        });
    }

    private int validValue(int originalValue) {
        if (originalValue < min) {
            return min;
        } else if (originalValue > max) {
            return max;
        } else {
            return originalValue;
        }
    }

}

File: jetbrains-core/src/software/aws/toolkits/jetbrains/ui/SliderPanel.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package software.aws.toolkits.jetbrains.ui

import com.intellij.openapi.ui.ValidationInfo
import org.jetbrains.annotations.TestOnly
import software.aws.toolkits.resources.message
import java.awt.event.FocusAdapter
import java.awt.event.FocusEvent
import javax.swing.JPanel
import javax.swing.JSlider
import javax.swing.JTextField

// A panel with a slider and text field, of which the slider and text field always synced up.
class SliderPanel(
    private val min: Int,
    private val max: Int,
    defaultValue: Int = min,
    minTick: Int = min,
    maxTick: Int = max,
    minorTick: Int = (max - min) / 30,
    majorTick: Int = (max - min) / 5,
    snap: Boolean = false
) {
    @Suppress("UnusedPrivateMember") // root element must be bound
    private lateinit var content: JPanel
    lateinit var slider: JSlider
        private set
    lateinit var textField: JTextField
        private set

    init {
        slider.majorTickSpacing = majorTick
        slider.minorTickSpacing = minorTick
        slider.maximum = maxTick
        slider.minimum = minTick
        slider.paintLabels = true
        slider.snapToTicks = snap
        slider.value = defaultValue
        slider.addChangeListener { textField.text = validValue(slider.value).toString() }
        textField.text = slider.value.toString()
        textField.addFocusListener(
            object : FocusAdapter() {
                // When the text field lost focus, we force the value to be valid to reset to
                // - default value if the input is not a valid integer, or
                // - min if it is smaller than min, or
                // - max if it is bigger than max.
                override fun focusLost(e: FocusEvent) {
                    val value = validValue(textField.text.toInt())
                    slider.value = value
                    textField.text = value.toString()
                }
            }
        )
    }

    var value: Int
        get() = slider.value
        set(value) {
            slider.value = value
        }

    fun validate(): ValidationInfo? {
        val value = textField.text.toIntOrNull()
        return if (value == null || value < min || value > max) {
            ValidationInfo(message("lambda.slider_validation", min, max), textField)
        } else {
            null
        }
    }

    fun setEnabled(enabled: Boolean) {
        slider.isEnabled = enabled
        textField.isEnabled = enabled
    }

    @get:TestOnly
    val isVisible: Boolean
        get() = slider.parent.isVisible && slider.isVisible && textField.isVisible

    private fun validValue(originalValue: Int): Int = originalValue.coerceAtMost(max).coerceAtLeast(min)
}
